librz/util/subprocess.c

```diff
@@ -1281,17 +1417,98 @@ RZ_API void rz_subprocess_free(RzSubprocess *proc) {
 	rz_strbuf_fini(&proc->err);
 	rz_sys_pipe_close(proc->killpipe[0]);
 	rz_sys_pipe_close(proc->killpipe[1]);
-	if (proc->stdin_fd != -1) {
+
+	if (proc->master_fd != -1) {
+		rz_sys_pipe_close(proc->master_fd);
+	}
+	if (proc->stdin_fd != -1 && proc->stdin_fd != proc->master_fd) {
 		rz_sys_pipe_close(proc->stdin_fd);
 	}
-	if (proc->stdout_fd != -1) {
+	if (proc->stdout_fd != -1 && proc->stdout_fd != proc->master_fd) {
 		rz_sys_pipe_close(proc->stdout_fd);
 	}
-	if (proc->stderr_fd != -1 && proc->stderr_fd != proc->stdout_fd) {
+	if (proc->stderr_fd != -1 && proc->stderr_fd != proc->stdout_fd && proc->stderr_fd != proc->master_fd) {
 		rz_sys_pipe_close(proc->stderr_fd);
 	}
 	free(proc);
 }
+
+/**
+ * \brief Call openpty(3) with the provided arguments
+ *
+ * \param slave_name The name of the slave PTY is stored in this
+ * This is marked as RZ_BORROW, so it's ownership no longer stays with the caller
+ * and is now owned by the returned RzPty struct
+ *
+ * \param term_params Terminal attributes (struct termios) for the forked process
+ * \param win_params Window attributes (struct winsize) for the forked process
+ *
+ * \return RzPty*
+ */
+RZ_API RZ_OWN RzPty *rz_subprocess_openpty(RZ_BORROW RZ_NULLABLE char *slave_name, RZ_NULLABLE void /* const struct termios */ *term_params, RZ_NULLABLE void /* const struct winsize */ *win_params) {
+	RzPty *pty = RZ_NEW0(RzPty);
+	int ret = rz_sys_openpty(&pty->master_fd, &pty->slave_fd, slave_name, NULL, NULL);
+
+	if (ret == -1) {
+		perror("openpty");
+		RZ_FREE(pty);
+		return NULL;
+	}
+
+	return pty;
+}
+
+/**
+ * \brief Call login_tty(3) on the provided \p pty
+ *
+ * \param pty RzPty struct
+ * \return bool true if login_tty succeeded, false otherwise
+ */
+RZ_API bool rz_subprocess_login_tty(RZ_BORROW RZ_NONNULL const RzPty *pty) {
+	rz_return_val_if_fail(pty, false);
+
+	int ret = rz_sys_login_tty(pty->slave_fd);
+	if (ret == -1) {
+		perror("login_tty");
+		return false;
+	}
+
+	return true;
+}
+
+/**
+ * \brief Closes the file descriptors associated with \p pty
+ *
+ * \param pty RzPty struct
+ * \return void
+ *
+ * No need to call this after you've used the \p pty in `rz_subprocess_start_opt`,
+ * since the file descriptors would already have been correctly closed
+ */
+RZ_API void rz_subprocess_close_pty(RZ_BORROW RZ_NONNULL const RzPty *pty) {
+	if (close(pty->master_fd) == -1) {
+		perror("close");
+	}
+	if (close(pty->slave_fd) == -1) {
+		perror("close");
+	}
+}
+
+/**
```

Comment:
This Doxygen is unnecessary and can be removed

Reply:
Agreed, but I'll let it stay.